File: laravel/app/Models/Cards/DeckCard.php
<?php

namespace App\Models\Cards;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;

class DeckCard extends Model
{
    protected $table = 'deck_cards';

    protected $fillable = ['quantity', 'is_commander', 'deck_id', 'card_id'];

    protected $casts = [
        'is_commander' => 'boolean',
    ];

    public function deck(): BelongsTo
    {
        return $this->belongsTo(Deck::class, 'deck_id');
    }

    public function card(): BelongsTo
    {
        return $this->belongsTo(Card::class, 'card_id');
    }

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!(($this->quantity === null || ($this->quantity >= 1 && $this->quantity <= 4)))) {
            $errors['quantity_range'] = 'A deck can contain between 1 and 4 copies of a card';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->is_commander === true && !($this->quantity === 1)) {
            throw new \RuntimeException('Commander card must appear exactly once in the deck');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function incrementAction($amount): void
    {
        // TODO: implement increment
    }

    public function decrementAction($amount): void
    {
        // TODO: implement decrement
    }

}

File: rails/app/models/deck_card.rb
class DeckCard < ApplicationRecord
  self.table_name = 'deck_cards'

  belongs_to :deck, class_name: 'Deck', inverse_of: :deck_cards
  belongs_to :card, class_name: 'Card', inverse_of: :deck_cards

  validates :quantity, numericality: { greater_than_or_equal_to: 1, less_than_or_equal_to: 4 }
  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:quantity_range, 'A deck can contain between 1 and 4 copies of a card') unless ((quantity.nil? || (quantity >= 1 && quantity <= 4)))
  end

  # Domain invariants — IMPLIES rules
  validate :validate_implies

  def validate_implies
    errors.add(:base, 'Commander card must appear exactly once in the deck') if (is_commander == true) && !(quantity == 1)
  end

  def to_s
    quantity.to_s
  end

  # Business operations

  def increment(amount)
    # TODO: implement increment
  end

  def decrement(amount)
    # TODO: implement decrement
  end
end
